Switch CI to GitHub Actions

smortex opened the pull request:
Adjust Rubocop configuration to match other choria projects and update the code so that we follow the style conventions we decided.

Setup GitHub Actions to replace CircleCI.

This PR is intended to check the behavior of the setup and the CI configuration might be extracted to be shared with other projects.

smortex commented:
I merged this branch in `master` on my fork so that the GitHub Action is triggered and it seems to do the right thing™:
https://github.com/smortex/package-agent/actions/runs/2885453071

ripienaar commented on .rubocop.yml:
```diff
@@ -1,3 +1,7 @@
+require:
```
this is copied from mco repo or where? Ideally we sync the style between that repo and these (I think I did that when i got them from Puppet Labs)

smortex replied:
Yes it was copied from choria-io/mcorpc-ruby-support:

```diff
--- /home/romain/Projects/choria-io/mcorpc-ruby-support/.rubocop.yml	2022-02-10 11:18:38.005071305 -1000
+++ .rubocop.yml	2023-04-04 13:30:07.403371049 -1000
@@ -1,8 +1,10 @@
+require:
+  - rubocop-performance
+  - rubocop-rake
+
 AllCops:
-  TargetRubyVersion: "2.4"
   DisplayCopNames: true
   DisplayStyleGuide: true
-  SuggestExtensions: false
   Exclude:
     - "module/**/*"
     - "vendor/**/*"
@@ -107,7 +109,7 @@
   Enabled: false
 
 Layout/LineLength:
-  Max: 180
+  Max: 200
 
 Metrics/PerceivedComplexity:
   Enabled: false
```

ripienaar commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,45 @@
+      - main
+      - master
+
+jobs:
```
once this work we should add it to choria-io/actions repo and just call that

smortex commented:
I completely forgot about this, and after installing a nighttly puppet 8 which bundle Ruby 3.2 I got hit by some regression this PR was fixing (`File#exists?` instead of `File#exist?`).

I rebased on top of master and added ruby 3.2 to the CI matrix.

smortex commented:
Ugh!  Ruby 3.2 support is broken and a lot of gems are need to be updated to unbreak it (rake, rspec, etc).  This leads us to https://github.com/ripienaar/mcollective-test/ which have not been touched in 10 years :sweat_smile: 

Does it make sense to move this gem under the choria-plugins / choria-io organization?

ripienaar commented:
Oh boy :) that gem uses rspec2 and iirc when I tried  moving it to 3 5 or so years ago it was essentially impossible. This is going to hurt. And yeah we should move it assuming we can make it work at all

smortex commented:
> Oh boy :)

Yeah :sob: 

I reverted the Ruby 3.2 changes so that this PR is up-to-date on top of master and works as expected: https://github.com/smortex/package-agent/actions/runs/4633956211

Once released, this will unbreak the module with AIO Puppet 8 (`File#exists?` -> `File#exist?`), but will lack Ruby 3.2 tests.

> This is going to hurt. And yeah we should move it assuming we can make it work at all

Once merged, I will focus on this task in a new dedicated PR.

ripienaar commented:
Thanks I have merged that other gem PR. Do you need a release of tjst to continue with this?

smortex commented:
Thanks!  I think we can at least merge this PR as it is now.  I installed a nightly version of Puppet at $WORK to see what breaks, and this PR should fix this module, so a new release would help but is not utterly required right now (I fixed the file on my single test machine) :-)

Extending the test suite to test against ruby 3.2 will require a lot of work and will not be quickly done I am afraid.  Having this merged will allow GitHub actions to tests the work in progress.

ripienaar commented:
Good stuff 
<img width="682" alt="Screenshot 2023-04-07 at 17 20 34" src="https://user-images.githubusercontent.com/82342/230624646-b2082141-fb0e-411e-b5d2-cab41029e399.png">
